Elliptic curve

abh2k opened the pull request:
revives #2449 

<!-- BEGIN RELEASE NOTES -->
* ntheory
  * implemented  elliptic curve 
<!-- END RELEASE NOTES -->

shikil committed:
First commit

Support 'in' operation

Calculate points in FF

shikil committed:
Add doctest

Add references

Fix discriminant

Add docstring

Implement part of torsion list

Return sympy type

Generate all torsion points

Fix type conversion

shikil committed:
Remove _char

shikil committed:
Add docstring

Docstring of add and mul

shikil committed:
Add minimal Weierstrass equation and j-invariant calculation

Fix type error

Calculate point order.

Implement elliptic curve point class.

Support negative point multiplier

shikil committed:
Support right multiplication

Fix j-invariant representation

shikil committed:
Calculate point addition in general Weierstrass equation.

test case

abhinav28071999 commented:
Can someone help me to resolve this please.

jksuom commented:
It looks like you should merge current master (https://github.com/sympy/sympy/wiki/Development-workflow#how-to-merge) before adding new commits. The last one should be removed first (`git reset HEAD^`). (If that does not work, you can remove the branch: `git branch -D elliptio-curve` and fetch it again.)

abhinav28071999 commented:
now its okay, i guess.

sylee957 commented:
Are you going to make the object an instance of `Basic`?
If not, I'd name more trivially like `EllipticCurveSolver` or because I'd expect naming issues like #18672 can arise in the future, if someone tries to introduce a proper 'Basic' elliptic curve later.

I think that it would better to make the object `Basic`, and drop much of the problematic domain stuff like `QQ` but fix it to use `Rational`.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,360 @@
+
+    """
+
+    def __init__(self, a4, a6, a1=0, a2=0, a3=0, domain=QQ):
```
It is probably efficient to use non-Basic domains internally but it is not necessary to have them as input parameters. I would suggest using `modulus=0` instead. Then domain would be `QQ` if `modulus` is 0 and `FF(modulus)` otherwise.

abhinav28071999 replied:
okay.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,360 @@
+from sympy.abc import x, y
```
These should not be global variables. They should be defined locally by something like `x, y = symbols('x y')` if they are needed.

abhinav28071999 commented:
@jksuom I read about elliptic curves and found that there is a lot of difference in algorithms when the domain is QQ or ,ZZ or is finite. So i suggest making difference class for elliptic curves over finite fields and rational fields.

jksuom commented:
The domain should be a field so `ZZ` need not be considered. The same code will work over almost all fields, different classes are necessary only in characteristics 2 and 3. Those are not important for applications in number theory so I would not include them now to save time. (Classes for them can be added later if someone is interested.)

abh2k committed:
corrected points function

minor changes

corrected some doctest

added call function

 corrected test cases

minor change

normalise

 corrected test case

changed domain

correctd torsion point

  minor change

  minor change

  minor change

  minor change

   added points_x and imporved code for order of points

  minor      improvements

corrected equation

corrected  test case

minor improvement

rename

changed  name and minor improvement

  removed         init

abhinav28071999 commented:
i have corrected the points function. I think that most of the things implemented was wrong. I will correct those.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -1,15 +1,16 @@
-from .residue_ntheory import sqrt_mod
+from .residue_ntheory import quadratic_congruence
 
+x, y = symbols('x y')
```
These symbols should not be global. It they are needed in some class or function, they should be defined only there.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -32,28 +33,29 @@ class EllipticCurve():
-        self._a3 = self._domain(a3)
-        self._a4 = self._domain(a4)
-        self._a6 = self._domain(a6)
+        self._a1 = sympify(self._domain(a1))
```
These private coefficients should not be sympified. Working with coefficients belonging to the domain is more efficient.

If this is made a subclass of Basic, then the coefficients should be sympified and made public: `self.a1`, ... But there is no need to do that yet.

abhinav28071999 replied:
FF(7)(5) returns `2 mod7`. But these are required in many operations and it gives error that this is not an integer. `as_int(FF(7)(5))` gives error.

jksuom replied:
Why should you want to use `as_int` on elements of a finite field?

abhinav28071999 replied:
It is used by other functions. Like  when calculating points, we call `quadratic congruence`, there it checks with `as_int`.  I suppose we could sympify  just before calling `quadratic congruence`

jksuom replied:
>  we could sympify just before calling quadratic congruence

Yes, we can sympify the arguments.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -32,28 +33,29 @@ class EllipticCurve():
-        self._domain = domain
+    def __init__(self, a4, a6, a1=0, a2=0, a3=0, modulus = 0):
+        if modulus == 0:
+            self._domain = QQ
```
I would write `domain = QQ` or `domain = FF(modulus)`, and then `self._domain = domain`. Then those lines below where coefficients are converted: `self._a1 = self._domain(a1)` could be written `self._a1 = domain(a1)`, ...

sylee957 commented:
SymPy's notation of finite field is wrong, and it may actually be called a 'field' if the order is prime
#13886

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,371 @@
+        return EllipticCurvePoint(0, 1, 0, curve)
+
+    def __init__(self, x, y, z, curve):
+        self.x = x
```
The coordinates have to be domain elements.

    dom = curve._domain
    self.x = dom(x)
    etc.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,371 @@
+            return p
+        if p.z == 0:
+            return self
+        x1, y1 = self.x, self.y
```
It is expected that `z = 1` in the formulas. Hence the coordinates have to be normalized.

    x1, y1 = self.x/self.z, self.y/self.z
    etc.

jksuom commented:
The first task is to make doctests pass. I have added some comments to that end.

abhinav28071999 commented:
@jksuom  I have a doubt. When adding point is done, when domain is `QQ` we have to perform a `1/QQ(num)` and when finite field is used then we will  need `mod_inverse(FF(modulus)(num), modulus)`. But  due to #18703  and #18706 , how we will perform.

jksuom commented:
Division `1/a` should work both in `QQ` and in a finite field.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -284,17 +284,18 @@ def point_at_infinity(curve):
             return self
-        x1, y1 = self.x, self.y
+        x1, y1 = self.x/self.z, self.y/self.z
         x2, y2 = p.x, p.y
```
Here as well.

abhinav28071999 commented:
`e = EllipticCurve(2,3 , modulus = 97 )`
`p = e(3,6)`
`p+p`
it returns (-17, 10) which is correct but it should be in  the domain.
I tried `dom(x3), dom(y3)` before returning point addition, but didn't work.

abhinav28071999 commented:
@jksuom  is there any trick to get through #18703 because the errors are probably due to that.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,373 @@
+            yint = (-x1**3 + a4*x1 + 2*a6 - a3*y1) / (a1*x1 + a3 + 2*y1)
+        x3 = slope**2 + a1*slope - a2 - x1 - x2
+        y3 = -(slope + a1) * x3 - yint - a3
+        return EllipticCurvePoint(dom(x3), dom(y3), 1, self._curve)
```
What is the purpose of `dom`?

jksuom commented:
The arguments `dom=QQ` should be integers. That is not true on line 316 in general.

jksuom commented:
You could find `z3 = lcm(x3.q, y3.q)` and then return the integer coordinates `x3*z3, y3*z3, z3` thus removing the denominators. Alternatively, you could make `__init__` handle rational coordinates properly by converting `p/q` to `QQ(p, q)`.

abhinav28071999 commented:
I think when `p = point(1/2,1,1)` is done then `1/2` gets converted to 0.5. So when the domain is `QQ`
then the input to the point should be expected to be `p = point(QQ(1,2 ), 1, 1)' if a fraction is required. With this it can be resolved. And the return type will be `QQ(p, q)` of the add when domain is add.

abhinav28071999 commented:
@jksuom Is there anyway to tell the domain of a variable.
I mean if `a = QQ(2,9)`. Then is there anything like this `a.domain` and this will return QQ.

jksuom commented:
>  is there any trick to get through #18703

It seems that `QQ()` is inconvenien for conversions because its arguments should be integers (although `gmpy` can handle also other numbers). It is better to use the general purpose `QQ.convert()`. So `_init_` could be

    conv = curve._domain.convert
    self.x = conv(x)
    etc.

It should then not be necessary to do conversions elsewhere.

jksuom commented:
>  is there anything like this a.domain and this will return QQ.

There are two implementations of `QQ`. The elements of one implementation are `gmpy` objects of type `mpq`. Those objects have no knowledge of SymPy and `QQ`.

abhinav28071999 commented:
@jksuom This should be working now. As stated earlier when the characteristic function is not  `2 or 3` then the elliptic curve can be reduces with some transformations `y' -> F(x, y)` and `x' -> F(x, y)`. I read some books on elliptic curve and almost everywhere i found the assumption that the elliptic curve is in its minimal form `y**2 = x**3 + a*x + b`. We could also implement something like this. We can make `self.y'` and `self.x'` as transformations in `__init__` of the curve and work on `y'**2 = x'**3 + a*x' + b` every where and before returning any output we can just `reverse transform` them.
This will make it easier to implement and verify. For example, i didn't found the algorithm used for addition of points for general curve anywhere, so i can't say that the algorithm used here is 100% correct.

jksuom commented:
If characteristic is different from 2 and 3, then the equation of an elliptic curve can be transformed into the form `y**2 = x**3 + a*x + b`. This is known as the Weierstrass normal form. It should not be hard to verify the addition formulas for the general case where `a1`, `a2` and `a3` do not vanish, but I think that it is not necessary for now. The Weierstrass form should suffice for applications to primality testing and factorization of integers.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,374 @@
+        self._b4 = 2 * a4 + a1 * a3
+        self._b6 = a3**2 + 4 * a6
+        self._b8 = a1**2 * a6 + 4 * a2 * a6 - a1 * a3 * a4 + a2 * a3**2 - a4**2
+        self._discrim = domain(-self._b2**2 * self._b8 - 8 * self._b4**3 - 27 * self._b6**2 + 9 * self._b2 * self._b4 * self._b6)
```
It would be best to use `conv (= domain.convert)` also in this `__init__`. We have seen that `domain()` is fragile.

jksuom replied:
```suggestion
       self._discrim = -b2**2 * b8 - 8*b4**3 - 27*b6**2 + 9*b2 * b4 * b6
```

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,374 @@
+from sympy import symbols
```
This could be `from sympy.core.symbol import symbols` and placed after `core.relational`.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,374 @@
+from sympy.core.relational import Eq
+from sympy.polys.domains import FiniteField, QQ, RationalField, FF
+from sympy.solvers.solvers import solve
+from sympy import sympify
```
Is this still needed?

abhinav28071999 commented:
`FF(97).to_sympy(118)` return 118. Is this correct?

jksuom commented:
The argument of `FF(97).to_sympy()` should be an object of `FF(97)`.

abhinav28071999 commented:
all tests have passed.  What is the next step?

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,375 @@
+        self.x = dom(x)
+        self.y = dom(y)
+        self.z = dom(z)
+        self._curve = curve
```
It should be checked somewhere that  that `(x, y, z)` lies on the curve. Perhaps this would be a suitable place. There could also be a function that would find the points with given `x` (and `z = 1`). That would be useful for finding examples.

abhinav28071999 replied:
Done.

jksuom commented:
This looks useful already but I would like to polish the code a bit. The expressions look better if the a-coefficients are converted in the beginning. Then the b-coefficients would automatically be in the same domain. I would also make the equation homogeneous by adding the z-generator. It might also be convenient to write the equation as a Poly instance (difference of lhs and rhs) as that would include the generators. (There would be no need to store x, y, z in the curve.)

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,396 @@
+        else:
+            domain = FF(modulus)
+        self._domain = domain
+        domain = domain.convert
```
We could convert the coefficients as the very first thing.

    a1, a2, a3, a4, a6 = map(domain.convert, (a1, a2, a3, a4, a6))

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,396 @@
+        self._b6 = a3**2 + 4 * a6
+        self._b8 = a1**2 * a6 + 4 * a2 * a6 - a1 * a3 * a4 + a2 * a3**2 - a4**2
+        self._discrim = domain(-self._b2**2 * self._b8 - 8 * self._b4**3 - 27 * self._b6**2 + 9 * self._b2 * self._b4 * self._b6)
+        self._a1 = domain(a1)
```
```suggestion
        self._a1 = a1
```
and similarly below.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,396 @@
+        self._a3 = domain(a3)
+        self._a4 = domain(a4)
+        self._a6 = domain(a6)
+        self.x, self.y = symbols('x y')
```
```suggestion
        self.x, self.y, self.z = x, y, z = symbols('x y z')
```

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,396 @@
+        self._a4 = domain(a4)
+        self._a6 = domain(a6)
+        self.x, self.y = symbols('x y')
+        self._eq = Eq(self.y**2 + self._a1*self.x*self.y + self._a3*self.y, self.x**3 + self._a2*self.x**2 + self._a4*self.x + self._a6)
```
```suggestion
        self._eq = Eq(y**2*z + a1*x*y*z + a3*y*z**2, x**3 + a2*x**2*z + a4*x*z**2 + a6*z**3)
```
It might also be possible to replace this by a Poly.

jksuom commented:
See also [this](https://github.com/sympy/sympy/pull/18682/files#r383251012).

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -37,22 +37,25 @@ def __init__(self, a4, a6, a1=0, a2=0, a3=0, modulus = 0):
+        self._a6 = a6
+        x, y, z = symbols('x y z')
+        self.x, self.y, self.z = x, y, z
+        self._eq = Eq(y**2*z + a1*x*y*z + a3*y*z, x**3*z + a2*x**2*z + a4*x*z + a6*z)
```
z is added to make this homogeneous of degree 3. So `a3*y*z**2`, `x**3 + a2*x**2*z + a4*x*z**2 + a6*z**3`.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -37,22 +37,25 @@ def __init__(self, a4, a6, a1=0, a2=0, a3=0, modulus = 0):
             domain = FF(modulus)
+        a1, a2, a3, a4, a6 = map(domain.convert, (a1, a2, a3, a4, a6))
         self._domain = domain
         domain = domain.convert
```
I don't think `domain` would be needed any more.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -37,22 +37,25 @@ def __init__(self, a4, a6, a1=0, a2=0, a3=0, modulus = 0):
+        b6 = a3**2 + 4 * a6
+        b8 = a1**2 * a6 + 4 * a2 * a6 - a1 * a3 * a4 + a2 * a3**2 - a4**2
+        self._b2, self._b4, self._b6, self._b8 = b2, b4, b6, b8
+        self._discrim = domain(-b2**2 * b8 - 8 * b4**3 - 27 * b6**2 + 9 * b2 * b4 * b6)
```
`domain` can be omitted as the b-constants are domain elements.

jksuom commented on sympy/ntheory/ec.py:
```diff
@@ -0,0 +1,399 @@
+        return '({}, {})'.format(self.x, self.y)
+
+    def __sub__(self, other):
+        return self + -other
```
This will probably be the same as `return self.__add__(-other)`.

jksuom commented:
I have a few minor comments but otherwise I think that this is ready be merged. Only the name should be changed. `ec.py` is not informative, I would suggest `elliptic_curves.py` (`git mv ...`).

abhinav28071999 commented:
@jksuom  Done.

abhinav28071999 commented:
how to remove this circular import error?

abhinav28071999 commented:
@jksuom I am not able to push the changes it is giving this error.


![Untitled](https://user-images.githubusercontent.com/41710346/75267130-cd4da300-581b-11ea-818b-021d80cbe6fa.png)

abhinav28071999 commented:
It got push, don't know why it was showing error earlier.

abhinav28071999 commented:
when i added elliptic curve in `ntheory\__init__.py` it is giving circular import error.

abhinav28071999 commented:
when i added elliptic curve in `ntheory\__init__.py` it is giving circular import error.

jksuom commented:
I wonder how that did happen. Only the module name was changed.

jksuom commented:
It looks like GitHub had some issues: https://twitter.com/githubstatus?lang=en

It is probably not necessary to import elliptic curves from `__init__.py` at this point.

abh2k committed:
changed  file name

jksuom commented:
It looks like GitHub had some issues: https://twitter.com/githubstatus?lang=en

It is probably not necessary to import elliptic curves from `__init__.py` at this point.

abhinav28071999 commented:
should i squash  all the commits.

jksuom commented:
We should preserve the commits of the original author, or at least some of them. Can you squash selected commits?

abhinav28071999 commented:
okay.

abhinav28071999 commented:
Done. Is this okay?

jksuom commented:
I'll look more closely tomorrow but I think it is ok.

jksuom commented:
Thanks, this should be a good basis for computations with elliptic curves. It can be extended later if desired. The `rank` method may be too hard to implement (I don't know of any algorithm) but it should be harmless to leave the stub alone.

I have removed WIP from the title.

abhinav28071999 commented:
@jksuom  Hey, I have made the proposal for quadratic sieve and Elliptic curve factorization. [Here](https://docs.google.com/document/d/1y_kLou7NwodV75puX88Lihdkx63sw5aQEc1zVuh8_NM/edit?usp=sharing) is the link. I have posted this on mailing list but there are too many posts coming lately so i though that this might skip. Please go through it and suggest any changes.
Thank you
